src/common/utilities.cpp

```diff
@@ -0,0 +1,495 @@
+    for (auto i = 0; i < dimensions.size() - 1; ++i) {
+        const auto& dim = dimensions.at(i);
+        chunk_strides.push_back(chunk_strides.back() *
+                                zarr::common::chunks_along_dimension(dim));
+        CHECK(chunk_strides.back());
+    }
+
+    // get chunk indices
+    std::vector<size_t> chunk_lattice_indices;
+    for (auto i = 0; i < dimensions.size() - 1; ++i) {
+        chunk_lattice_indices.push_back(chunk_idx % chunk_strides.at(i + 1) /
+                                        chunk_strides.at(i));
+    }
+    chunk_lattice_indices.push_back(chunk_idx / chunk_strides.back());
+
+    // make shard lattice indices
+    std::vector<size_t> shard_lattice_indices;
+    for (auto i = 0; i < dimensions.size(); ++i) {
+        shard_lattice_indices.push_back(chunk_lattice_indices.at(i) /
+                                        dimensions.at(i).shard_size_chunks);
+    }
+
+    std::vector<size_t> chunk_internal_strides(1, 1);
+    for (auto i = 0; i < dimensions.size() - 1; ++i) {
+        const auto& dim = dimensions.at(i);
+        chunk_internal_strides.push_back(chunk_internal_strides.back() *
+                                         dim.shard_size_chunks);
+    }
+
+    size_t index = 0;
+
+    for (auto i = 0; i < dimensions.size(); ++i) {
+        index +=
+          (chunk_lattice_indices.at(i) % dimensions.at(i).shard_size_chunks) *
+          chunk_internal_strides.at(i);
+    }
+
+    return index;
+}
+
+size_t
+common::bytes_per_chunk(const std::vector<Dimension>& dimensions,
+                        const SampleType& type)
+{
+    auto n_bytes = bytes_of_type(type);
+    for (const auto& d : dimensions) {
+        n_bytes *= d.chunk_size_px;
+    }
+
+    return n_bytes;
+}
+
+const char*
+common::sample_type_to_dtype(SampleType t)
+
+{
+    static const char* table[] = { "u1", "u2", "i1", "i2",
+                                   "f4", "u2", "u2", "u2" };
+    if (t < countof(table)) {
+        return table[t];
+    } else {
+        throw std::runtime_error("Invalid sample type.");
+    }
+}
+
+const char*
+common::sample_type_to_string(SampleType t) noexcept
+{
+    static const char* table[] = { "u8",  "u16", "i8",  "i16",
+                                   "f32", "u16", "u16", "u16" };
+    if (t < countof(table)) {
+        return table[t];
+    } else {
+        return "unrecognized pixel type";
+    }
+}
+
+size_t
+common::align_up(size_t n, size_t align)
+{
+    EXPECT(align > 0, "Alignment must be greater than zero.");
+    return align * ((n + align - 1) / align);
+}
+
+std::vector<std::string>
+common::split_uri(const std::string& uri)
+{
+    const char delim = '/';
+
+    size_t begin = 0;
+    auto end = uri.find_first_of(delim);
+
+    std::vector<std::string> out;
+    while (end <= std::string::npos) {
+        if (end > begin) {
+            out.emplace_back(uri.substr(begin, end - begin));
+        }
+
+        if (end == std::string::npos) {
```

Comment:
Is this check necessary? If `end == std::string::pos` the loop will be terminated anyway, right?

Reply:
No, not sure what that's doing there.